### PR TITLE
Allow trailing commas on same-line. Enforce await in async.

### PR DESCRIPTION
**Pull request by jdcumpson**
WISOTT

@MatthewHerbst and I discussed this, the comma-dangle thing we have decided is a matter of taste, I am voting for more leniency as the more common use-case is destructuring objects like
```javascript
const { a, b, } = myObj; 
```
Rather than needing new-lines to keep the comma-dangle contract in-tact
```javascript
const {
  a,
  b,
} = myObj;
```
---
The `async/await` change is more nuanced. While both functions execute synchronously as a developer I would expect a function using the `async` directive to actually be asynchronous, and therefore execute on a different call stack of the event loop. Enforcing this behaviour would make debugging easier for developers to understand, as well as ensure that synchronicity bugs are less prevalent because you cannot accidentally execute synchronous code before other synchronous code. The example is presented below:

Define some set of intentionally asynchronous methods
```javascript
async function doFirstThing() { console.log('Start first thing') }

// note the await in `doSecondThing`
async function doSecondThing() { await console.log('Start second thing') }
```
and imagine having a function in which executes them all at the same time, followed by some other synchronous function.
```javascript
function executeAllThings() {
  const promise1 = doFirstThing();
  const promise2 = doSecondThing();
  console.log('Started all async tasks');
}
```
#### Example 1
While there is no guarantees in the JavaScript interpreter to ensure that `doFirstThing` or `doSecondThing` execute *later*, since they are marked with `async` one would expect that they output in the following order **on the call-stack** e.g.
```
Call stack 1:
  - doFirstThing: enqueue an async action (inside it's function block), same as setTimeout(..., 0)
  - doSecondThing: enqueue an async action, same as doFirstThing
  - execute console.log('Started all async tasks');
Call stack 2:
  - execute doFirstThing function block
  - execute doSecondThing function block
```

#### Example 2
However, since `doFirstThing` does not use `await` the behaviour actually changes implicitly to 
```
Call stack 1:
  - doFirstThing: no await, execute console.log('Start first thing');
  - doSecondThing: *with* await, enqueue function block as previous example
  - execute console.log('Started all async tasks');
Call stack 2:
  - execute doSecondThing
```

The second example it will not be obvious **why** the code executes this way when both functions are `async`. The proposed change will enforce all functions to use `await` to ensure they implicitly behave the same.

The outputs can be compared too for concrete execution comparison:
```
Without await enforced:                               With await enforced:
console.log('Start first thing')                    console.log('Started all async tasks')
console.log('Started all async tasks')          console.log('Started first thing')
console.log('Started second thing')             console.log('Started second thing')
```

The way to way to do it without writing `setTimeout` everywhere? Use `await` on any value assignment, as it works for any value type, not just *thenable* types.

```javascript
async function doFirstThing() {
  await console.log('Start first thing');
}
```
or in a variable assignment of a non-async value, for example
```javascript
async function doFirstThing() {
  const a = await 'a';
}
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -33,7 +33,7 @@ module.exports = {
     'block-scoped-var': 0,
     'brace-style': [2, '1tbs', { allowSingleLine: false }],
     'camelcase': [2, { properties: 'never' }],
-    'comma-dangle': [2, 'always-multiline'],
+    'comma-dangle': [2, 'always'],
```
Comment (MatthewHerbst) [inline index.js:36]: What's the desire for this?

File: index.js [hunk 2]
```diff
@@ -180,7 +180,7 @@ module.exports = {
     'prefer-spread': 2,
     'quotes': [2, 'single', 'avoid-escape'],
     'radix': 2,
-    'require-await': 0,
+    'require-await': 2,
```
Comment (MatthewHerbst) [inline index.js:183]: I'm not sure I agree with this rule. For example, our GraphQL codebase marks all async functions with `async`, even if we don't use `await` in them, simply as a way of making clear to developers that it will return a `Promise`

**Comment by jdcumpson** [comment]
I also forgot to mention that in the `comma-dangle` change, we lose protection from accidental sparse array initialization like
```javascript
const array = [1, ,];  //  [1, undefined]
```
But I do not think that this as useful as keeping comma-dangle contract for single and multiline.

**Comment by MatthewHerbst** [comment]
cc @steverice FYI for the `require-await` change proposal

File: index.js (hunk 1 above)
Comment (steverice) [reply index.js:36]: I agree with more leniency, but this does the opposite. Now there are _fewer_ ways to write correct code that will pass the linter, and because the rule was previously `always-multiline`, every single-line statement that has been using these rules will now be invalid.

I'm not sure that this was the intent. The tin says "allow" but the code says "require".
This change would _also_ have the effect of breaking any pre-ECMAScript 2017 code, as requiring trailing commas in function calls in those versions makes a functional difference.

Perhaps you want something closer to
```javascript
"comma-dangle": ["error", {
    "arrays": "always-multiline",
    "objects": "ignore",
    "imports": "always-multiline",
    "exports": "always-multiline",
    "functions": "always-multiline"
}]
```
but even then this change would remove the check that multiline objects have comma dangle which, IMO, is a regression. Multiline dangle makes for cleaner diffs.

The ideal solution to accomplish this would be to modify the linter with an option that requires on multiline and ignores on single line. Not sure how you (or ESLint) feel about that 🙂 

However, to @MatthewHerbst 's point, it's unclear to me what functionality this allows that previously was not. With or without the comma dangle, your destructuring example is equivalent:
```javascript
const data = { a: '1', b: '2', c: '3' };
const { a, b } = data;
console.log(a, b);
```
`> 1 2`
```javascript
const data = { a: '1', b: '2', c: '3' };
const { a, b, } = data;
console.log(a, b);
```
`> 1 2`

File: index.js (hunk 2 above)
Comment (steverice) [reply index.js:183]: I also don't think this works, though not because it conflicts with current GraphQL style as @MatthewHerbst suggests (if that were the case, we'd just disable the rule in that codebase).

Rather, I agree with the goal, but don't think this accomplishes it.

Allow me to expand on your examples a bit:
```javascript
// this function passes the rule
async function doFirstThings() {
  console.log('Start first thing');
  await console.log('Start second thing');
  console.log('Start third thing');
}

// this function passes as well
async function doFourthThing() {
  await console.log('Start fourth thing');
}
```

So what happens when we run `executeAllThings()`, calling the two methods as you've defined it?
```
Call stack 1:
  - doFirstThings line 1: no await, execute console.log('Start first thing');
  - doFirstThings line 2: *with* await, yield execution until this promise resolves
  - doFirstThings line 3: not executed at all on this stack
  - doFourthThing: *with* await, yield execution until this promise resolves
  - execute console.log('Started all async tasks');
Call stack 2:
  - execute doFirstThings line 2
  - execution of doFirstThings resumes at line 3
Call stack 3:
  - execute doFourthThing
```

So you've set up `doFirstThings` line 2 and `doFourthThing` in a race to resolve. The output is:
```
Start first thing
Start second thing
Start fourth thing
Started all async tasks
Starts third thing
```

I don't think that's better. It certainly doesn't seem intuitive that things would execute in this order.

Again, if the goal is as you describe, I believe what you want is a different rule. Namely, something that would ensure all calls to an `async` function are `await`ed or themselves wrapped in a `Promise`. i.e. if we change the definition of `executeAllThings` to:
```javascript
async function executeAllThings() {
  const promise1 = await doFirstThings();
  const promise2 = await doFourthThing();
  console.log('Started all async tasks');
}
```

then we get an output much more in line with what we might expect:
```
Start first thing // this executed synchronously
Start second thing // starting here, all following calls are async
Start third thing
Start fourth thing
Started all async tasks
```

Even removing the `await` from one of the `do(First|Fourth)Thing(s)` methods, thus making it invalid according to this rule, we get the same results.

If we don't care about the execution order of `doFirstThings` vs. `doFourthThings`, we can wrap them in a `Promise.all` and `await` that, but still have predictability over how it relates to the synchronous code in the function.

*tl;dr*
Essentially, this change suggests that missing `await` inside an `async` function is indicative of a mistake; I assert that missing `await` (or a `Promise` wrapper when _calling_ an `async` function is where the mistakes lie.

**Review by steverice — CHANGES_REQUESTED** [review]
See my comments for the details. I think in both cases, however, these rules do not accomplish what you are looking for and, in fact, no currently-available ESLint configuration (that I can find) does.

File: index.js (hunk 2 above)
Comment (jdcumpson) [reply index.js:183]: You're not wrong about the execution order, however the synchronization of the promises is not in scope for this change. As you said if you are concerned about synchronization you would use a `Promise.all` or another method that can handle the synchronizing of promises.

The point is that if you omit `await` in the `doFirstThing()` you get the behaviour that is completely incongruent with when the final/resolved value from an `async` method is returned. In this case, it is not about the initial execution, in fact you could still have a function with `await` inside of it that looks like
```
async doFirstThing() {
  console.log('Start first thing');
  const output = await someOtherThing();
  return other thing;
}
```
And then the call stack would still look like you described above **but** the resulting call would be guaranteed to callback only after the next callstack executes. Therefore maintaining consistency of execution among all `async` functions. 

If there is never *any* `await` used inside the function block you will return the result of the function synchronously, and in doing-so may create synchronization issues that are implicitly hard to reason about when debugging - because the functions using `async` ought to be run **later** which means at least one call-stack in the future. Otherwise the `async` directive ought not be used.

*tl;dr*
It's not about synchronizing the execution of the callbacks, it's standardizing that they all will return at least tick in the future, and never execute synchronously if they have `async` on the function.

File: index.js (hunk 1 above)
Comment (jdcumpson) [reply index.js:36]: I'm not sure I understand what you mean here:
> but even then this change would remove the check that multiline objects have comma dangle which, IMO, is a regression. Multiline dangle makes for cleaner diffs.

The desired effect is enforcing comma-dangle on both multi-line and single-line. I do agree that I did not mean for it to be for functions though - I will update that. I think my solution should reflect
```javascript
"comma-dangle": ["error", {
    "arrays": "always",
    "objects": "always",
    "imports": "always",
    "exports": "always",
    "functions": "ignore"
}]
```

Since eslint actively ignores functions now as per https://github.com/eslint/eslint/blob/master/lib/rules/comma-dangle.js#L49

Again, I think this may be a matter of taste, since I think that the diff argument is not valid.

Comment (MatthewHerbst) [reply index.js:36]: If it's a matter of taste, I think we should aim for more leniency, and then be stricter with local rules within individual repos. That way we make adoption/continued use of this config as easy as possible for downstream users.

**Comment by jdcumpson** [comment]
As discussed, I don't think this rule would actually prevent developers from making the mistake I am trying to prevent. Although I believe it would help for our internal web repository - I do no think our yield function runtimes behave the same across all of our platforms. I'm going to close this and probably open this change for repos that use the babel environment in which yield statements guarantee next tick semantics. 

TL;DR - neither of us think it's a bad rule, but it would be more trouble than it's worth applying it globally. Perhaps to bring up again in the future, for now use in each project override as needed.
